test/README.md

```diff
@@ -0,0 +1,84 @@
+Candid test data
+================
+
+This directory contains Candid test data, to be used to check Candid
+implementations for compliance.
+
+This test suite covers
+
+ * parsing Candid service descriptions
+ * given a Candid type, decoding binary Candid data, asserting success or failure
+ * given a Candid type, parsing textual Candid data, asserting success or failure
+ * comparing such decoded/parsed values, using the Host language’s notion of equality.
+
+Because this test data necessarily has to be host-language agnostic, it is not
+possible to
+
+ * assert the decoded value, as it would be host-language specific
+ * cover encoding, as it would require specifying a host-language value,
+   and also because Candid encoding is not canonical
+
+Implementers are advised to extend this test suite with randomzied
+round-tripping tests (host value → candid → host value).
+
+Format
+======
+
+The test suite contains these kind of files:
+
+ * `<name>.good.did`:
+   A Candid service description, with valid candid syntax.
+   Implementations should be able to parse them.
+
+ * `<name>.bad.did`:
+   A Candid service description, with invalid candid syntax.
+   Implementations should reject parsing them.
+
+ * `<name>.test.did`:
+   A set of Candid tests, written in a grammar that extends the Candid type and value grammar:
+
+   The root production is
+   ```
+   <enctest>  ::= <def>;* <test>;*
+   ```
+   to allow any number of type definitions followed by encoding tests with grammar
+   ```
+   <test> ::=
+      | assert <input> : <tuptype> <desc>
+      | assert <input> !: <tuptype> <desc>
+      | assert <input> == <input> : <tuptype> <desc>
+      | assert <input> != <input> : <tuptype> <desc>
+   <input> ::= <text> | blob <text>
+   <desc> ::=  <text>?
+   ```
+   where the four forms assert
+    * that the input can be decoded/parsed at that type
+    * that the input cannot be decoded/parsed at that type
+    * that the two inputs values decode/parse to the same value
+    * that the two inputs values decode/parse to different values.
```

Comment:
So this assertion fails if they don't parse/decode? That actually makes sense, but is worth calling out.

Reply:
Good point, done